### DISCUSSION TITLE
How to insert with selecting foreign keys

### DISCUSSION THREAD

**Discussion by ISibboI** [diesel-rs/diesel#3726] [Q&A] [answered]
I have three tables with `REFERENCES` constraints as follows:

```sql
CREATE TABLE languages (
	id SERIAL PRIMARY KEY,
	english_name TEXT NOT NULL UNIQUE
);
CREATE TABLE word_types (
	id SERIAL PRIMARY KEY,
	english_name TEXT NOT NULL UNIQUE
);
CREATE TABLE words (
	word TEXT NOT NULL,
	word_type INTEGER REFERENCES word_types,
	language INTEGER REFERENCES languages,
	PRIMARY KEY(word, word_type, language)
);
```

I would like to insert into words with a single database round trip, under the assumption that `languages` and `word_types` have already been populated. In the postgres CLI, I simply write the following query, and it works like a charm:

```sql
INSERT INTO words (word, word_type, language) VALUES (
    'toy',
    (SELECT id FROM word_types WHERE english_name = 'Noun'),
    (SELECT id FROM languages WHERE english_name = 'English')
);
```

Whenever the selected word type and language exists, the insert just works. And when either the word type or the language does not exist, then the query produces an error as expected. However, I am lost when trying to transform this into diesel. I made the following attempt:

```rust
diesel::insert_into(words::table).values((
    words::word.eq(word.word),
    words::language.eq(languages::table.select(languages::id)
        .filter(languages::english_name.eq(&word.language_english_name))
        .single_value()),
    words::word_type.eq(word_types::table.select(word_types::id)
        .filter(word_types::english_name.eq(&word.word_type))
        .single_value()),
)).execute(connection)?;
```

However this has the same problem as in #1918. The inner `SELECT` statements may return nothing, and for SQL this equals setting `words.language` or `words.word_type` to `NULL`, but they are both marked as `NOT NULL`. In diesel, that means that both `words::language.eq` and `words::word_type.eq` expect an `Integer`, but the result of `single_value` is `Nullable<Integer>`.

In #1918, there was a workaround, and then there is also a [comment](https://github.com/diesel-rs/diesel/issues/1918#issuecomment-917409911) stating that this workaround is not required anymore since #2820 was merged, providing an `assume_not_null` function.

However, I am lost when trying to make above diesel query work with `assume_not_null`. It seems like I can only call it on `languages::id` or `word_types::id`, but it seems to have no effect.

How can I make the query above compile with using `assume_not_null`?

**Comment by weiznich** [comment, marked as answer]
Using `assume_not_null` might be the right solution here if you are sure that your subquery will always return a value. You need to call `assume_not_null()` on the subquery, which means the following query should work:

```rust
diesel::insert_into(words::table).values((
    words::word.eq(word.word),
    words::language.eq(languages::table.select(languages::id)
        .filter(languages::english_name.eq(&word.language_english_name))
        .single_value().assume_not_null()),
    words::word_type.eq(word_types::table.select(word_types::id)
        .filter(word_types::english_name.eq(&word.word_type))
        .single_value().assume_not_null()),
)).execute(connection)?;
```

If one of the subqueries returns no value the insert should fail with an constraint error.

**Reply by ISibboI** [reply]
Great thanks, that worked!
